File: app/src/main/java/com/aso/senote/data/source/local/dao/NoteDAO.java
/*
 * Copyright (c) ASO 2020.
 */

package com.aso.senote.data.source.local.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.aso.senote.data.models.Note;

import java.util.List;

@Dao
public abstract class NoteDAO {

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    abstract long insertNote(Note note);

    @Update
    abstract void updateNote(Note note);

    @Delete
    public abstract void deleteNotes(Note... note);

    @Transaction
    public void upsert(Note note) {
        long id = insertNote(note);
        if (id == -1)
            updateNote(note);
    }

    @Query("SELECT * from note_table WHERE notebookId = :NotebookId ORDER BY date_created DESC")
    public abstract LiveData<List<Note>> getNotesForNotebook(long NotebookId);

    @Query("SELECT * from note_table WHERE noteId = :noteId")
    public abstract LiveData<Note> getNoteById(long noteId);

    @Query("SELECT * from note_table ORDER BY date_created DESC")
    public abstract LiveData<List<Note>> getAllNotesByDateCreated();

    @Query("DELETE FROM note_table")
    public abstract void deleteAllNotes();
}

File: app/src/main/java/com/aso/senote/data/source/local/dao/NoteDAO.kt
/*
 * Copyright (c) ASO 2020.
 */
package com.aso.senote.data.source.local.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import com.aso.senote.data.models.Note

@Dao
abstract class NoteDAO {
    @Insert(onConflict = OnConflictStrategy.IGNORE)
    abstract fun insertNote(note: Note): Long

    @Update
    abstract fun updateNote(note: Note)

    @Delete
    abstract fun deleteNotes(vararg note: Note)

    @Transaction
    open fun upsert(note: Note): Long {
        val id = insertNote(note)
        if (id == -1L) updateNote(note)
        return id
    }

    @Query("SELECT * from note_table WHERE notebookId = :NotebookId ORDER BY date_created DESC")
    abstract fun getNotesForNotebook(NotebookId: Long): LiveData<List<Note>?>

    @Query("SELECT * from note_table WHERE noteId = :noteId")
    abstract fun getNoteById(noteId: Long): LiveData<Note?>

    @Query("SELECT * from note_table WHERE noteId = :noteId")
    abstract fun getNote(noteId: Long): Note?

    @get:Query("SELECT * from note_table ORDER BY date_created DESC")
    abstract val allNotesByDateCreated: LiveData<List<Note>?>

    @Query("DELETE FROM note_table")
    abstract fun deleteAllNotes()
}